File: server/DigitalTwins.DAL/Entities/Widget.cs
using DigitalTwins.Common.Enums;

namespace DigitalTwins.DAL.Entities;

public class Widget
{
    public long Id { get; set; }

    public string Title { get; set; } = string.Empty;

    public WidgetType Type { get; set; }

    public long? DatastreamId { get; set; }
    
    public long TemplateId { get; set; }

    public Datastream Datastream { get; set; } = null!;

    public Template Template { get; set; } = null!;
    
    public IEnumerable<WidgetDevice> WidgetDevices { get; set; } = null!;
}

File: client/src/types/widget.ts
import { Datastream } from '@@types/datastream';
import { Template } from '@@types/template';

export enum WidgetType {
  Toggle,
}

export interface Widget {
  id: number;
  title: string;
  type: WidgetType;
  datastreamId: Datastream['id'];
  templateId: Template['id'];
  value: 0 | 1;
}

export type UpdateWidgetArgs = Omit<Pick, 'title' | 'datastreamId'> & {
  widgetId: Widget['id'];
};
